dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/AbstractAwsApiTask.java

```diff
@@ -0,0 +1,95 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.edwmigration.dumper.application.dumper.connector.redshift;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.auth.profile.ProfileCredentialsProvider;
+import com.amazonaws.services.redshift.AmazonRedshift;
+import com.amazonaws.services.redshift.AmazonRedshiftClient;
+import com.google.common.io.ByteSink;
+import com.google.edwmigration.dumper.application.dumper.ConnectorArguments;
+import com.google.edwmigration.dumper.application.dumper.task.AbstractTask;
+import com.google.edwmigration.dumper.plugin.ext.jdk.progress.RecordProgressMonitor;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVPrinter;
+
+/** Abstract class that provides methods for connecting with AWS API and writing results. */
+public abstract class AbstractAwsApiTask extends AbstractTask<Void> {
+
+  AWSCredentialsProvider credentialsProvider;
+  Class<? extends Enum<?>> headerEnum;
+  Optional<AmazonRedshift> redshiftClient;
+
+  public AbstractAwsApiTask(
+      AWSCredentialsProvider credentialsProvider,
+      String zipEntryName,
+      Class<? extends Enum<?>> headerEnum) {
+    super(zipEntryName);
+    this.headerEnum = headerEnum;
+    this.redshiftClient = Optional.empty();
+    this.credentialsProvider = credentialsProvider;
+  }
+
+  @Nonnull
+  public AbstractAwsApiTask withRedshiftApiClient(AmazonRedshift redshiftClient) {
+    this.redshiftClient = Optional.of(redshiftClient);
+    return this;
+  }
+
+  public AmazonRedshift redshiftApiClient() {
+    return redshiftClient.orElseGet(
+        () -> AmazonRedshiftClient.builder().withCredentials(credentialsProvider).build());
+  }
+
+  public Void writeRecordsCsv(@Nonnull ByteSink sink, Stream<Object[]> records) throws IOException {
+    CSVFormat format = FORMAT.builder().setHeader(headerEnum).build();
+    try (RecordProgressMonitor monitor = new RecordProgressMonitor(getName());
+        Writer writer = sink.asCharSink(StandardCharsets.UTF_8).openBufferedStream()) {
+      CSVPrinter printer = format.print(writer);
+
+      for (Object[] record : records.collect(toImmutableList())) {
+        monitor.count();
+        printer.printRecord(record);
+      }
+    }
+    return null;
+  }
+
+  public static Optional<AWSCredentialsProvider> createCredentialsProvider(
+      ConnectorArguments arguments) {
+    if (arguments.getIAMProfile() != null) {
```

Comment:
The calls to the methods in `arguments` are repeated. The code could be simplified to:
```
return Optional.ofNullable(arguments.getIAMProfile())
  .map(ProfileCredentialsProvider::new)
  .or(() -> 
    Optional.ofNullable(arguments.getIAMAccessKeyID())
      .flatMap(accessKeyId ->
        Optional.ofNullable(arguments.getIAMSecretAccessKey())
          .map(secretAccessKey ->
            new AWSStaticCredentialsProvider(
              new BasicAWSCredentials(accessKeyId, secretAccessKey)))));
```

Reply:
What do you think about simply assigning arguments to local variables? You code looks nice but I think it's more clear with simple if statements. In the future we may need to extend it with additional Providers. Adding next if statement is easy, chaining Optionals may be confusing for the reader.

Reply:
Using local variables with ifs also works. Possibly, the second part of the method could be further simplified:

```
return optionalIf(iamAccessKey != null && iamSecretAccessKey != null,
    () -> new AWSStaticCredentialsProvider(
        new BasicAWSCredentials(iamAccessKey, iamSecretAccessKey)));
```

Reply:
I implemented similar idea inline because `optionalIf` method is defined in TeradataUtils class and we shouldn't import it. Later we can move it to some shared utils package.